morf-testsupport/src/main/java/org/alfasoftware/morf/testing/UpgradeTestHelper.java

```diff
@@ -244,22 +274,30 @@ private boolean jiraIdIsValid(final String jiraId) {
    * Turn the list of classes into a list of objects.
    */
   private List<UpgradeStep> instantiateAndValidateUpgradeSteps(Iterable<Class<? extends UpgradeStep>> stepClasses) {
-    return Streams.stream(stepClasses)
-        .map(stepClass -> {
-          UpgradeStep upgradeStep;
-          try {
-            Constructor<? extends UpgradeStep> constructor = stepClass.getDeclaredConstructor();
-            // Permit package-protected classes
-            constructor.setAccessible(true);
-            upgradeStep = constructor.newInstance();
-          } catch (Exception e) {
-            throw new RuntimeException(e);
-          }
-
-          validateUpgradeStep(upgradeStep);
-
-          return upgradeStep;
-        })
-      .collect(Collectors.toList());
+
+    List<String> errors = new ArrayList<>();
+    List<UpgradeStep> upgradeSteps = new ArrayList<>();
+
+    for (Class<? extends UpgradeStep> stepClass : stepClasses) {
+      UpgradeStep upgradeStep;
+      try {
+        Constructor<? extends UpgradeStep> constructor = stepClass.getDeclaredConstructor();
+        // Permit package-protected classes
+        constructor.setAccessible(true);
+        upgradeStep = constructor.newInstance();
+      }
+      catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+
+      errors.addAll(validateUpgradeStep(upgradeStep));
+      upgradeSteps.add(upgradeStep);
+    }
+
+    if (!errors.isEmpty()) {
+      fail(String.format("Upgrade step errors were found:%n%s%n", String.join(System.lineSeparator(), errors)));
+    }
+
+    return upgradeSteps;

```

Comment:
This returns a mutable ArrayList.
Maybe use ImmutableList instead.
Or wrap this in Collections.unmodifiableList(list)